core/domain/suggestion_services.py

```diff
@@ -348,3 +348,36 @@ def reject_suggestion(suggestion, reviewer_id, review_message):
     feedback_services.create_message(
         get_thread_id_from_suggestion_id(suggestion.suggestion_id), reviewer_id,
         feedback_models.STATUS_CHOICES_IGNORED, None, review_message)
+
+
+def get_all_suggestions():
+    """Gets all the suggestions.
+
+    Returns: list(Suggestion). All the suggestions.
+    """
+    return [get_suggestion_from_model(s)
+            for s in suggestion_models.GeneralSuggestionModel.get_all()]
+
+
+def query_suggestions(queries):
+    """Queries for suggestions.
+
+    Args:
+        queries: list(tuple). A list of queries. The first element in each tuple
+            is the field to be queried, and the second element is its value.
+
+    Returns:
+        list(Suggestion). A list of suggestions that match the given query
+            values.
+    """
+    suggestions = get_all_suggestions()
```

Comment:
This is too much stuff. Better to use an appropriately-formatted datastore query.

Define the exact filters that are allowed and make sure to validate them.

Also read up on the limits of datastore querying.  You'll be limited in what combinations of filters you can support, and will need to define appropriate error handling.

Finally there should be a controller that feeds into this, and backend tests for that (which ensure that the controller is translating requests to the right queries and that the correct results are being returned). There should be tests for each type of filter and various combinations of filters.

Reply:
@nithusha21  ah, retrieving all the data and then filtering is never a good idea. I thought maybe there is some way to construct the query dynamically based on the number of filters. I fear that we will have to stick to having different services for different types of queries. It might be worth to only implement the ones which will actually be important in the beginning.

Reply:
Hmmm. I'll try dynamically constructing the query (if possible). If not possible let's drop this idea :)

Reply:
It should be possible -- just need to make sure to define the allowed fields, as well as limits on how complex the query is allowed to be. This is the sort of thing that may be worth capturing in a (brief) design doc...

Reply:
By how "complex" what do you mean? I planned to write this to match all the given filters (ANDed together). You can only query on certain fields though (will add a check for that). But any number of these fields is fine. Also instead of erroring, I just returned an empty list (would an error be better?)

Reply:
Read the App Engine documentation on queries and indexes. There are limits to how complex a query can be; more complex queries will throw an error.

Reply:
Oh ok. Thanks!